### PR TITLE
fix(config): fix order of precedence in option processing

### PR DESCRIPTION
**Pull request by larsbutler**
Default values were overriding values set at higher levels of
precedence, namely command line args (which have the highest
precedence!).

Patch includes specific tests cases, which could be expanded to include
ini and keyring sources of config options. (The tests only cover
defaults, env, and command line sources of config params.)

### CODE REVIEW

File: simpl/config.py [hunk 1]
```diff
@@ -477,8 +477,15 @@ def parse_cli(self, argv=None, permissive=False):
...
-            if kwargs.get('default') is None:
+            original_default = kwargs.get('default')
+            if original_default is not None:
                 kwargs['default'] = argparse.SUPPRESS
+                # Since we're suppressing defaults, we need to somehow preserve
+                # the default value for the help text:
+                help_text = kwargs.get('help', '')
+                kwargs['help'] = '%s (default: %s)' % (
+                    help_text, original_default
+                )
```
Comment (samstav) [inline simpl/config.py]: This changed from

```python
if kwargs.get('default') is None:
    kwargs['default'] = argparse.SUPPRESS
```

to

```python
if kwargs.get('default') is not None:
    kwargs['default'] = argparse.SUPPRESS
    # then add default and help text manually
```

What's going on here? First change I can think of is that default values of `None` will be included in the help/usage output now.

**Comment by samstav** [comment]
Nice catch @larsbutler!

Is there a test case for

> For example, if the defaults
> define a value for the option `foo`, and `args` contains `foo=None`,
> the `args` should override the default.

?

And I left a comment demonstrating my confusion about `argparse.SUPPRESS` and the help/usage output. Maybe you could include one test that asserts that the `--help` output includes/doesn't-include what you expect?

File: tests/test_config.py [hunk 2]
```diff
@@ -430,5 +430,63 @@ def test_default_help_formatter(self, mock_stdout):
 """
         self.assertEqual(mock_stdout.getvalue(), expected)
 
+
+class TestConfigPrecedence(unittest.TestCase):
```
Comment (ziadsawalha) [inline tests/test_config.py]: We've been missing these tests. Glad to see them come in...

File: simpl/config.py [hunk 3]
```diff
@@ -477,8 +477,15 @@ def parse_cli(self, argv=None, permissive=False):
         options = []
         for option in self._options:
             kwargs = option.kwargs.copy()
-            if kwargs.get('default') is None:
```
Comment (ziadsawalha) [inline simpl/config.py]: `argparse.SUPPRESS is not None` - so the deleted line handled `default=argparse.SUPPRESS` correctly, I think and the new one may miss it?

**Pull request merged**